DecipherNow/gm-fabric-dashboard
# Fabric wide dash

This PR is my attempt to rebase all of my previous `fabric-wide-dash` work with my more recent work bringing Go dashboard support up to parity with the JVM.

A significant achievement is that the dashboard can run either with or without a Fabric Server. 

When `index.html` looks like this, the dashboard works as before
```html
<meta property="fabricServer" content="__FABRIC_SERVER__">
```

However, when the endpoint for a Fabric Server is provided, all other `index.html` meta tags are ignored and the dashboard pulls services and metrics information from the Fabric Server
```html
<meta property="fabricServer" content="http://localhost:9990">
```

This should allow the dashboard to be deployed per-instance as before and also in the new manner where a single dashboard monitors an entire fabric. 

To test running against a Fabric server:
`npm run mock-sds` and then configure the fabric server meta tag

To test running directly against a Go microservice without a Fabric server:

```bash
docker pull drfogout/metricssimple
docker run -it -p 1337:10001 -p 20001:20001 -p 30001:30001 drfogout/metricssimple
```
 and then configure the runtime meta tag in `index.html`. Make sure the fabric server meta tag is not set.


Known Issues:
- Fabric Server mode does not yet load threads metrics from the Fabric Server when the threads component mounts
- The `mock-sds` mock service does not yet mock go metrics
- Groups and other aspects of the Fabric Server have not been implemented
- Per agreement with the backend engineers, instances is not included in the response to the GET /services route. The `mock-sds` endpoint will have to be modified in anticipation of these changes.
